WIP: Move Material Parameters from the MaterialSrg to a Material Parameter Struct

kh-huawei opened the pull request:
## What does this PR do?

This PR is related to the proposed [RFC 152: Support for multiple Material-Types in Raytracing Hit Shaders](https://github.com/o3de/sig-graphics-audio/issues/152)

This PR doesn't change anything for Raytracing yet:It is instead a first step to change existing material shader functions to expect a `MaterialParameter` - struct containing the relevant parameters instead of accessing 'global' Srg - variables. 

In short, this changes every access to e.g. `MaterialSrg::m_baseColor` to `params.m_baseColor` and provides everything necessary to facilitate that.

### MaterialParameter - Struct Layout

- New classes `MaterialShaderParameter` and `MaterialShaderParameterLayout` 
  - They act as a layer between the parameter data from the `.material` file (or the material instance) and the shader parameters, and we change the behaviour of the Material-System such that on the CPU side everything refers to an entry in the `MaterialShaderParameterLayout` (either by name or by a cached index) instead of the Input-Index in the Srg, so that we can redirect the parameters values into either a StructuredBuffer or to the SRG as before.
    - To maintain some backwards compatability we check if the SRG contains the parameter directly (e.g. `MaterialSrg::m_baseColor`), or search for it in a `m_params` - struct (e.g. `MaterialSrg::m_params.m_baseColor`)
  - To determine the layout of the `MaterialParameter` struct we could extract the layout of the SRG members, but this presents somewhat of a chicken and egg problem: we want to change the MaterialSrg so it contains only a buffer with the Parameters (bindless or otherwise), making it impossible to extract the layout in the first place. Therefore we decided to treat the parameters in the `.materialtype` as the only source of material parameters, and generate the layout by collecting all parameters with a shader connection during the first (abstract) processing step of the material-pipelines. We write the layout of the `MaterialParameter` struct into a `<name>_parameters.azsli` next to the generated material shader files and modify the shader templates to include that generated file before the MaterialSRG - files.
  - A struct can't hold `Texture2D`: We had to convert all Image-parameters to a read-index Parameter, and modify the shaders to get the texture from the BindlessSRG where neccesary. As a workarround for non-bindless support, we collect the Image-Views in an array we manage ourselves, similar to the Raytracing-Implementation.
  - One problem with this were the material functors: They can access SRG - members directly that are never described anywhere else in the `.materialtype` - definition. E.g. the `ConvertEmissiveUnit` - functor assumes a SRG member with the name `m_emissiveIntensity` exists in the MaterialSrg, and that name is converted to the correct `InputConstantIndex` in the constructor. To get arround this, we introduced the field `shaderparams` in the functor definition that contain the name and data-type of parameters referenced only by the functors, and are not present in the Material-Parameters otherwise.

- Global `MaterialInstanceHandlerInterface`
  - Creates and maintains the `SceneMaterialSrg` and the GPU-Buffers holding the parameters
  - Each `Material` - instance registers itself during construction, receiving a `MaterialTypeIndex`, a `MaterialInstanceIndex` and a `MaterialSrg` that is either the global `SceneMaterialSrg`, or a separate Material-Srg specific for that material-instance, depending on the SRG layout.
  - The `MaterialTypeIndex` and `MaterialInstanceIndex` are written to the `DrawSrg` when the `MeshDrawPacket` is created.
  - The `Compile` function is called once per frame from the `RHISystem`, which collects and uploads the changed Parameters from all registered materials to the GPU buffers (one Parameter-Buffer for each Material-Type)
  - The implementation is inside the `MaterialSystem.cpp/h`, and Material-Instances are currently global across all scenes. I'm not fully convinced this is the best place for this, but it was the most convenient. And there is at least one material that is created before any scene, so we'd need a fallback behaviour if we move this to e.g. a featureprocessor somewhere.
 
- Shader Changes
  - Add a `const MaterialParameter params` as a input parameter to almost every Material-shader function (e.g. `EvaluateVertexGeometry`).
  - Change the `VertexAndPixel` - Shaders of the Material-Pipeline to fetch the Material-Parameters from the `SceneMaterialSrg` using the function `GetMaterialParameters()`, which uses the `MaterialTypeIndex` and `MaterialInstanceIndex` from the Draw-Srg
  - Change all occurences of `MaterialSrg::<name>` to `params.<name>` for non-texture parameters
  - Change ocurrences of `MaterialSrg::<textureName>` to `asTexture(params.<textureName>)`, where the `asTexture()` function fetches a texture from the Bindless Srg.
  - Change access to the texture sampler from `MaterialSrg::m_sampler` to `GetMaterialTextureSampler()`, which simply returns the default sampler from the `SceneMaterialSrg`.
  - Most of the Input-Functions (e.g. `GetBaseColorInput()`) remain unchanged

- MaterialCanvas Changes
  - This was generally very straightforward: We changed the Shader-Templates for the `MaterialGraphName` the same way we changed the other shaders
  - Accessing a Parameter is defined in the `.materialgraphnode` - files
    - For general parameters we simply changed this from `MaterialSrg::inValue` to `params.inValue`, and it mostly worked as is
    - For Textures we changed this from `MaterialSrg::inValue.Sample(MaterialSrg::inSampler,...)` to `asTexture(params.inValue).Sample(GetMaterialTextureSampler(), ...)`
      - This does use the default material sampler, and ignores the Sampler Parameters provided in the Material Graph.  

- Backwards compatability
  - If a shader uses the material pipeline and wants to use material functions provided by the Engine, the shaders need to be changed to use the `SceneMaterialSrg` and the `MaterialParameters`
  - If a shader does not use the Material-Pipeline, it can't really use the `SceneMaterialSrg`. But if it still wants to use the material functions provided by the engine (e.g. `SilhouetteGather.azsl`), that shader can define a `MaterialParameter` struct manually that contains the neccesary parameters, and place it in a `m_params` - struct inside the SRG. The `MaterialInstanceHandler` recognizes that the material doesn't use the `SceneMaterialSrg` - layout and will create a separate MaterialSrg, and will forward the Parameter-Values to the right SRG members.
  - If a shader does neither use the Material-Pipeline nor the material-functions of the engine, this will work as before: Each material-Instance has it's own Material-SRG, and the Parameter-Values will be forwarded to the SRG-Members as described in the `.materialtype`.
  - If the Bindless-SRG is not supported, the shaders will include the `MaterialSrg`, which contains a `MaterialParameters m_params` member, and a `Texture2D m_textures[]` where we collect the referenced textures of that instance. The `asTexture()` function will then fetch the textures from that array.

msat-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialSystem.cpp:
```diff
@@ -38,8 +42,473 @@ namespace AZ
+            }
+            if (!m_sceneMaterialSrgShaderAsset)
+            {
+                // neither the VBufferMaterial nor the fallback - asset-Ids were valid: try again, maybe the assetProcessor is still running
```
vbuffer

kh-huawei replied:
removed the comment

msat-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialSystem.cpp:
```diff
@@ -38,8 +42,473 @@ namespace AZ
+            auto materialAsset = material->GetAsset();
+            auto materialTypeAsset = materialAsset->GetMaterialTypeAsset();
+
+            // Note: the Asset-Id of the Material-type isn't enough, i also need to check the Shader Options
```
Where do you check the shader options?

kh-huawei replied:
Nowhere, since this isn't needed yet. Updated the comment.

msat-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialTypeSourceData.cpp:
```diff
@@ -1035,7 +1055,70 @@ namespace AZ
+            MaterialShaderParameterLayout shaderParamsLayout{};
+            // the first entry is always the material type
+            shaderParamsLayout.AddMaterialParameter<uint32_t>("m_materialType", true);
+            // TODO: remove this (debug only)
```
Should this be removed before merging?

kh-huawei replied:
i'm actually verifying the instance id on the gpu, so i've changed the comment

msat-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialTypeAssetCreator.cpp:
```diff
@@ -348,45 +348,122 @@ namespace AZ
+            MaterialShaderParameterLayout shaderParamsLayout{};
+            // the first entry is always the material type
+            shaderParamsLayout.AddMaterialParameter<uint32_t>("m_materialType", true);
+            // TODO: remove this (debug only)
```
Should this be removed before merging?

kh-huawei replied:
see above: i'm actually checking the Material Instance ID in the shader now. Changed the comment

msat-huawei commented on Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialGraphCompiler.cpp:
```diff
@@ -141,7 +141,9 @@ namespace MaterialCanvas
-            BuildMaterialSrgForCurrentNode();
+            // Note: The MaterialSrg has been replaced with the MaterialParameters - struct, which is generated automatically from the
+            // MaterialProperties with a shaderInput connection
+            // BuildMaterialSrgForCurrentNode();
```
Do we still need the commented-out line in the future? If not please remove

kh-huawei replied:
unclear. I've added a `deprecated` comment to that function. but i don't think we should remove it just yet.

msat-huawei reviewed (commented):
Looks good! I still had some comments, but I think most of them should be easy to fix.

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/Materials/Eye/Eye_SurfaceEval.azsli:
```diff
@@ -168,76 +157,79 @@ Surface EvaluateSurface_Eye(
     
-    surface.diffuseAmbientOcclusion = GetOcclusionInput(MaterialSrg::m_diffuseOcclusionMap, MaterialSrg::m_sampler, uvs[MaterialSrg::m_diffuseOcclusionMapUvIndex], real(MaterialSrg::m_diffuseOcclusionFactor), o_diffuseOcclusion_useTexture, uvDxDy, customDerivatives);
-    surface.specularOcclusion = GetOcclusionInput(MaterialSrg::m_specularOcclusionMap, MaterialSrg::m_sampler, uvs[MaterialSrg::m_specularOcclusionMapUvIndex], real(MaterialSrg::m_specularOcclusionFactor), o_specularOcclusion_useTexture, uvDxDy, customDerivatives);
+    // These parameters aren't in the skin.materialtype, so they don't exist in the MaterialParameters - struct
```
Which params are not in skin.materialtype? Are we not handling AO for eyes?

kh-huawei replied:
Two issues here: the comment should state `eye.materialtype`, and the value should be set to 1.

But no, eyes don't support AO: These are defined in `OcclusionPropertyGroup.json`, which is not included in the materialtype, and therefore never exposed to the user, which means `GetOcclusionInput` would always return 1.

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/Materials/MultilayerPBR/StandardMultilayerPBR_Common.azsli:
```diff
@@ -8,10 +8,20 @@
 
 #pragma once 
 
+#define USE_BINDLESS_SRG 1
```
Enabled globally for all platforms

moudgils commented on Gems/Atom/Feature/Common/Assets/Shaders/Materials/MultilayerPBR/StandardMultilayerPBR_Common.azsli:
```diff
@@ -23,79 +33,6 @@
-
-ShaderResourceGroup MaterialSrg : SRG_PerMaterial
-{
-    Texture2D m_blendMaskTexture;
```
Where did this code get moved to? I had trouble finding it.

kh-huawei replied:
Same as for all materials: The `MaterialParameter` - struct is generated from the parameters in the `StandardMultiLayerPBR.materialtype` and placed in `[project]/Cache/Intermediate Assets/materials/types/standardmultilayerpbr_parameters.azsli`.

The multilayer materialtype includes `StandardMultiLayerPBR_LayerProperties.json` three times with a different `shaderInputsPrefix`, which eventually results in e.g. `m_layer1_m_baseColor`, `m_layer2_m_baseColor` and `m_layer3_m_baseColor`. The `m_blendMaskTexture` is also very straighforward referenced in the materialtype, and also ends up in the struct.

moudgils commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialInstanceHandler.h:
```diff
@@ -0,0 +1,58 @@
+            Data::Instance<MaterialShaderParameter> m_materialShaderParameter{ nullptr };
+        };
+
+        class IMaterialInstanceHandler
```
Add a comment to describe the purpose of the class

moudgils reviewed (commented):
Only done a partial review. Will continue going through it tomorrow.

kh-huawei commented on Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Materials/SceneMaterialSrg.azsli:
```diff
@@ -0,0 +1,44 @@
+MaterialParameters GetMaterialParameters()
+{
+    int materialTypeBufferIndex = SceneMaterialSrg::m_materialTypeBufferIndices[DrawSrg::m_materialTypeId];
+    ByteAddressBuffer materialInstances = Bindless::GetByteAddressBuffer(materialTypeBufferIndex);
```
The SceneMaterialSrg only works if the Bindless SRG is supported, and if that isn't, the single - MaterialSrg should be used.

I could refactor this to include two bound arrays (ByteAddressBuffers for the parameters and Texture2D for the textures), but at that point the workaround in the Bindless SRG should also work, making the refactor pointless?

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/PersistentIndexAllocator.h:
```diff
@@ -0,0 +1,91 @@
+    namespace RPI
+    {
+        template<typename T>
+        class PersistentIndexAllocator
```
added a comment

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Reflect/Material/MaterialFunctor.h:
```diff
@@ -328,6 +341,8 @@ namespace AZ
             //! Bit position uses MaterialPropertyIndex of the property.
             MaterialPropertyFlags m_materialPropertyDependencies;
+
+            AZStd::vector<MaterialFunctorShaderParameter> m_functorShaderParameters;
```
added a comment

kh-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialShaderParameterLayout.cpp:
```diff
@@ -0,0 +1,567 @@
+                    desc->m_structuredBufferBinding.m_elementCount,
+                    count);
+                AZ_Assert(
+                    desc->m_typeName == typeName,
```
Not entirely sure if a property can be defined multiple times, or if multiple properties can map to the same shader parameter.

Either case, both properties write to the same parameter, and whichever property is processed last wins...

msat-huawei replied:
I would at least create a warning if this happens

kh-huawei replied:
added a warning

kh-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialShaderParameterLayout.cpp:
```diff
@@ -0,0 +1,567 @@
+            // materialtype into a non-abstract materialtype.
+            // Shaders that do not use an abstract material-type (e.g. the SilhouetteGather - shader) can still use the (new) shader
+            // material functions by manually defining a MaterialParameter - struct. If that struct is part of the SRG with the name
+            // "m_params", we set the parameter values directly in the SRG.
```
agreed, but i'm not sure where to document this...

kh-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialSystem.cpp:
```diff
@@ -38,8 +42,473 @@ namespace AZ
+        {
+            if (!m_sceneMaterialSrgShaderAsset)
+            {
+                // Load the default vbuffer material asynchronously
```
fixed the comment.

kh-huawei reviewed (commented):
rebased and adressed several comments

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialShaderParameter.h:
```diff
@@ -0,0 +1,413 @@
+                }
+            }
+
+            bool SetParameterRaw(const MaterialShaderParameterLayout::Index& index, const AZStd::span<uint8_t>& data)
```
refactored the entire file a bit

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialShaderParameter.h:
```diff
@@ -0,0 +1,413 @@
+            AZStd::span<const uint8_t> GetRawBufferParameterData(
+                const MaterialShaderParameterDescriptor* desc, const uint32_t deviceIndex = 0) const;
+
+            uint64_t GetStructuredBufferDataSize() const
```
refactored the entire file a bit

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialShaderParameterLayout.h:
```diff
@@ -0,0 +1,129 @@
+
+            bool WriteMaterialParameterStructureAzsli(const AZ::IO::Path& filename) const;
+
+            AZStd::vector<MaterialShaderParameterDescriptor>& GetDescriptors()
```
no, removed.

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialShaderParameterLayout.h:
```diff
@@ -0,0 +1,129 @@
+                AZ_TYPE_INFO(MaterialShaderParameterDescriptor::BufferBinding, "{9D088B39-B392-4603-8465-94E48D6083C9}");
+                static void Reflect(ReflectContext* context);
+
+                size_t m_elementSize;
```
initialized

gadams3 commented on Gems/Atom/Tools/MaterialCanvas/Assets/MaterialCanvas/GraphData/Nodes/Textures/sample_texture_2d.materialgraphnode:
```diff
@@ -13,7 +13,7 @@
             ],
             "instructions": [
-                "SLOTTYPE(outColor) outColor = TransformColor4(MaterialSrg::inValue.Sample(MaterialSrg::inSampler, (float2)inUV), inColorSpaceFrom, inColorSpaceTo);"
+                "SLOTTYPE(outColor) outColor = TransformColor4(asTexture(params.inValue).Sample(GetMaterialTextureSampler(), (float2)inUV), inColorSpaceFrom, inColorSpaceTo);"
```
This is a breaking change. Material Canvas supports creating and customizing texture samplers, which get injected into the material SRG. If that is no longer generated then proper support needs to be added somewhere else for creating and customizing texture samplers. Users have existing content with different wrap and filter modes and other settings that they cannot create with the handful of PBR materials we provide.

kh-huawei replied:
added Texture Samplers State to MaterialProperties, and register them in a `m_samplers` - array in the srg

gadams3 reviewed (changes requested):
I've gone through most of the shader files and a lot of the code changes. Concern is that this breaks custom texture samlers. Being able to configure texture samplers, wrap modes, and filter modes et cetera was a long standing request that material canvas was compensating for. Samplers could be created and configured per texture, or shared between them with a couple of new nodes. Is there a plan or another mechanism for supporting this? It would be great if they could be included in the material type and added to the material prams as well.

kh-huawei commented:
> I've gone through most of the shader files and a lot of the code changes. Concern is that this breaks custom texture samplers. Being able to configure texture samplers, wrap modes, and filter modes et cetera was a long standing request that material canvas was compensating for. Samplers could be created and configured per texture, or shared between them with a couple of new nodes. Is there a plan or another mechanism for supporting this? It would be great if they could be included in the material type and added to the material prams as well.

I still have a few comments to get to, but in the meantime to give a brief update about the custom texture samplers:

I've added the SamplerState as a configurable Material Property in the Material System. I can't store the sampler state in the MaterialParameters  - struct directly though, since the sampler state can't really be stored i a structured buffer.

What I'm doing instead is adding a fixed-size array of samplers to each Material-SRG (shared or otherwise), and store an index for that array in the Material-Parameters. The samplers in that array will be configured (and deduplicated) by a registry, where each SamplerState - MaterialProperty registers itself.

In practice i now have 16 unique configurations of SamplerStates in the SceneMaterialSrg (shared between all materials) and 4 in the SingleMaterialSrg (for one material-instance), configured on demand from the  Material-Properties.
The sizes of these sampler-arrays are arbitrary, but i wanted to keep them small, since they are filled with samplers, whether they are used or not. This would also make it quite easy to add configurable samplers to e.g. the StandardPBR, but i didn't want to change any UI just yet.

The RPI - side of this was easier than expected; modifying the Material-Properties less so.

kh-huawei commented on Gems/Atom/Feature/Common/Assets/Materials/Types/EnhancedPBR.azsli:
```diff
@@ -49,9 +53,9 @@
     // Callback function for ParallaxMapping.azsli
-    DepthResult GetDepth(float2 uv, float2 uv_ddx, float2 uv_ddy)
+    DepthResult GetDepth(const MaterialParameters params, float2 uv, float2 uv_ddx, float2 uv_ddy)
     {
-        return SampleDepthFromHeightmap(MaterialSrg::m_heightmap, MaterialSrg::m_sampler, uv, uv_ddx, uv_ddy);
+        return SampleDepthFromHeightmap(GetMaterialTexture(params.m_heightmap), GetMaterialTextureSampler(), uv, uv_ddx, uv_ddy);
     }
 
     #include <Atom/Feature/Common/Assets/Shaders/Materials/EnhancedPBR/EnhancedPBR_VertexData.azsli>
```
done. Now only `ParallaxInput,azsli` knows about the Materialparameters.

kh-huawei commented on Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Materials/SceneMaterialSrgDefinition.azsli:
```diff
@@ -0,0 +1,27 @@
+{
+    Buffer<int> m_materialTypeBufferIndices;
+
+    Sampler m_sampler
```
Update: I've added samplers as a material property and made them configurable

msat-huawei commented on Gems/Atom/RHI/Code/Include/Atom/RHI.Reflect/SamplerState.h:
```diff
@@ -95,6 +95,9 @@ namespace AZ::RHI
         BorderColor m_borderColor = BorderColor::TransparentBlack;
+
+        bool operator==(const SamplerState& other) const { return GetHash() == other.GetHash(); }
+        bool operator!=(const SamplerState& other) const { return GetHash() != other.GetHash(); }
```
These comparisons ignore hash collisions. It's very unlikely that those will occur, but they can still happpen

msat-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/Material.cpp:
```diff
@@ -67,34 +67,39 @@ namespace AZ
+            }
+            else
+            {
+                return RHI::ResultCode::Fail;
```
Please add an assert or error here, so we can see why the `Init` failed.

kh-huawei replied:
added an assert

msat-huawei commented on Gems/Atom/RPI/Code/Source/RPI.Public/Material/MaterialShaderParameter.cpp:
```diff
@@ -0,0 +1,373 @@
+                //     ->GetDeviceShaderResourceGroup(deviceIndex)
+                //     ->GetData()
+                //     ->SetConstant(index, readIndex);
+                return m_shaderResourceGroup->SetConstant(*index, readIndex);
```
I assume this means that single material SRGs will not work with MultiGPU. Ideally this could be fixed in this PR, or alternatively create an issue for this.

kh-huawei replied:
created issue #18494

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialShaderParameter.h:
```diff
@@ -0,0 +1,413 @@
+                if (index && index->IsValid())
+                {
+                    // TODO: this doesn't work, i can't set different constant values in different device-SRGs (yet)
+                    // m_shaderResourceGroup->GetRHIShaderResourceGroup()
```
Created issue #18494 to keep track of that issue.

kh-huawei commented on Gems/Atom/Feature/Common/Assets/Materials/Pipelines/Common/DepthPass.azsli:
```diff
@@ -17,7 +17,7 @@ VsOutput VertexShader(VsInput IN, uint instanceId : SV_InstanceID)
     VsSystemValues SV;
     SV.m_instanceId = instanceId;
-    VsOutput OUT = EvaluateVertexGeometry(IN, SV);
+    VsOutput OUT = EvaluateVertexGeometry(IN, GetMaterialParameters(), SV);
```
moved the material to the last parameter

kh-huawei commented on Gems/Atom/Tools/MaterialCanvas/Assets/MaterialCanvas/GraphData/Nodes/MaterialOutputs/BasePBR/MaterialGraphName_Common.azsli:
```diff
@@ -39,14 +43,6 @@
 
 option bool o_normal_override_enabled = false;
 
-ShaderResourceGroup MaterialSrg : SRG_PerMaterial
-{
-    // m_placeHolder keeps MaterialSrg from being compiled out, causing SRG mismatches while assets are processing when iterating on graphs
-    bool m_placeHolder;
-
-    // O3DE_GENERATED_MATERIAL_SRG_BEGIN
-    // O3DE_GENERATED_MATERIAL_SRG_END
-}
```
see comment, added support for custom texture samplers

kh-huawei commented on Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialGraphCompiler.cpp:
```diff
@@ -141,7 +141,9 @@ namespace MaterialCanvas
             // this, all variables would be included in the SRG and material type. The shader compiler would eliminate unused variables from
             // the compiled shader code. The material type would fail to build if it referenced any of the eliminated variables.
-            BuildMaterialSrgForCurrentNode();
+            // Note: The MaterialSrg has been replaced with the MaterialParameters - struct, which is generated automatically from the
+            // MaterialProperties with a shaderInput connection
```
see comment, added support for custom texture samplers

kh-huawei commented on Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Materials/MaterialSrg.azsli:
```diff
@@ -0,0 +1,63 @@
+#ifndef USE_BINDLESS_SRG
+    // the MaterialParameters can only contain texture indices, and no Texture2D directly.
+    // if we aren't using the bindless SRG, we register the Textures here, and the index refers to the index in this array 
+    Texture2D m_textures[];
```
Added a workaround that uses one small, bound array for each material Instance.

kh-huawei commented on Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Materials/SceneMaterialSrg.azsli:
```diff
@@ -0,0 +1,65 @@
+
+#include <Atom/Features/SrgSemantics.azsli>
+
+// Note: This limit is arbitrary, but the samplers will exist in the SRG even if they are unused, so try to keep it low.
```
added a notice about the deduplication below

kh-huawei commented on Gems/Atom/Feature/Common/Assets/Materials/Types/StandardMultilayerPBR.azsli:
```diff
@@ -438,15 +443,16 @@ option enum class OpacityMode {Opaque, Cutout, Blended, TintedTransparent} o_opa
-            real depth = GetNormalizedDepth(startDepth, stopDepth, uvs[MaterialSrg::m_parallaxUvIndex], real2(0,0), real2(0,0));
+            real startDepth = real(-params.m_displacementMax);
+            real stopDepth = real(-params.m_displacementMin);
+            // the heightmap and sampler doesn't matter here, since they will simply be passed on to the GetDepth() - function of the multilayer material, which will 
```
completed comment

kh-huawei commented on Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Materials/SceneMaterialSrg.azsli:
```diff
@@ -0,0 +1,65 @@
+    return materialInstances.Load<MaterialParameters>(sizeof(MaterialParameters) * DrawSrg::m_materialInstanceId);
+}
+
+sampler GetMaterialTextureSampler(const int samplerIndex = 0)
```
added Comments

gadams3 commented on Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialGraphCompiler.cpp:
```diff
@@ -1176,8 +1179,7 @@ namespace MaterialCanvas
             {
                 // Sampler states are currently not configurable and will not be added added to the material type, just the material SRG.
-                if (!materialPropertyValueSlot || materialPropertyValueSlot->GetValue().empty() ||
-                    materialPropertyValueSlot->GetValue().is<AZ::RHI::SamplerState>())
+                if (!materialPropertyValueSlot || materialPropertyValueSlot->GetValue().empty())
                 {
```
The comments need to be updated but this should be fantastic. Have you tried creating a material type and changing sampler states from the material editor or material component? This should allow for dynamic sampler state changes which was not supported before.

kh-huawei commented:
I've squashed the commits and rebased them, and explicitly tested modifying sampler-states from the material-instance, and it works! I had to introduce some reference counting for the sampler states, since without it each change in the UI would result in a new sampler config that blocks a slot.

![Animation](https://github.com/user-attachments/assets/ee5caec8-b966-433a-976e-15f3c567133d)

kh-huawei commented:
i've changed the target-branch to 'material_refactoring', so we can get partial reviews and merges for all changes

jhmueller-huawei commented on Gems/Atom/Feature/Common/Code/Source/Material/SubsurfaceTransmissionParameterFunctor.cpp:
```diff
@@ -83,9 +83,44 @@ namespace AZ
+            context.GetMaterialShaderParameter()->SetParameter(m_transmissionTintThickness, transmissionTintThickness);
+        }
+
+        bool SubsurfaceTransmissionParameterFunctor::UpdateShaderParameterConnections(const RPI::MaterialShaderParameterLayout* layout)
+        {
+            bool valid = true;
+            if (m_scatterDistance.ValidateOrFindIndex(layout) == false)
+            {
+                AZ_Error(
+                    "SubsurfaceTransmissionParameterFunctorSourceData",
+                    false,
+                    "Could not find shader parameter '%s'",
+                    m_scatterDistance.GetName().GetCStr());
+                valid &= false;
+            }
+
+            if (m_transmissionParams.ValidateOrFindIndex(layout) == false)
+            {
+                AZ_Error(
+                    "SubsurfaceTransmissionParameterFunctorSourceData",
+                    false,
+                    "Could not find shader parameter '%s'",
+                    m_transmissionParams.GetName().GetCStr());
+                valid &= false;
+            }
+
+            if (m_transmissionTintThickness.ValidateOrFindIndex(layout) == false)
+            {
+                AZ_Error(
+                    "SubsurfaceTransmissionParameterFunctorSourceData",
+                    false,
+                    "Could not find shader parameter '%s'",
+                    m_transmissionTintThickness.GetName().GetCStr());
+                valid &= false;
+            }
+            return valid;
```
Is the goal here to try all and report all errors (as implemented) or would it be better to fail fast (return false immediately)? Saw this pattern multiple times.

kh-huawei replied:
The goal is to report all errors, since changes that fix these errors usually cause a rebuild of many (if not all) material-shaders, which can take several minutes.
The ability to see and fix several very similar errors all at once quickly became very desirable during these refactorings.

moudgils reviewed (commented):
Overall great changes and a much needed refactor. I have added some more comments. Nothing too big but given the considerable changes related to the material system are you able to add info related to all the testing done as part of this work. For example - testing related to existing materials within AutomatedTesting +ASV, using material editor, using material canvas, using MC generated materials, testing across different platforms, test with AZ_TRAIT_FORCE_SINGLE_MATERIAL_SRG enabled and without, add perf related impact of this work. The point is that we want to be very confident that these changes do not introduce breaking or new issues.

kh-huawei commented on Gems/Atom/RPI/Code/CMakeLists.txt:
```diff
@@ -204,6 +205,7 @@ if(PAL_TRAIT_BUILD_TESTS_SUPPORTED AND PAL_TRAIT_BUILD_HOST_TOOLS)
                 Gem::${gem_name}.Public
                 Gem::${gem_name}.TestUtils
     )
+    target_compile_options(${gem_name}.Tests PRIVATE -O0)
```
removed the line. 
It disables optimization to allow better debugging without having to compile the entire engine in debug-mode, but it should never have been checked in.

kh-huawei commented on Gems/Atom/RPI/Code/Source/Platform/Mac/Atom_RPI_Traits_Mac.h:
```diff
@@ -8,3 +8,6 @@
 #define AZ_TRAIT_MASKED_OCCLUSION_CULLING_SUPPORTED 0
+
+// Note: This requires both AZ_TRAIT_FORCE_SINGLE_MATERIAL_SRG and AZ_TRAIT_MAX_TEXTURES_PER_MATERIAL in AzslcHeader.azsli
+#define AZ_TRAIT_REGISTER_TEXTURES_PER_MATERIAL
```
the android header defines explicit bindless support, so i assumed this works. I have yet to test that though.

kh-huawei replied:
i've enabled single-material mode and non-bindless for android by default for now.

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialSystem.h:
```diff
@@ -8,23 +8,117 @@
+            {
+                bool m_valid = false;
+                // Note: The material either uses the SceneMaterialSrg, which is shared between all materials, or it uses a separate SRG for
+                // each instance. We don't share anything only between instances of the same MaterialType.
```
clarified the comment a bit.

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/PersistentIndexAllocator.h:
```diff
@@ -0,0 +1,98 @@
+    namespace RPI
+    {
+        // Manages indices for an persistent array where the contents never move, but the array can contain holes
+        // if an entry in the middle was released. The holes will be filled with new entries if possible.
```
added thread-safety warning

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/TextureSamplerRegistry.h:
```diff
@@ -0,0 +1,38 @@
+    namespace RPI
+    {
+
+        class TextureSamplerRegistry
```
added a comment

kh-huawei commented on Gems/Atom/Feature/Common/Code/Source/Material/ConvertEmissiveUnitFunctor.cpp:
```diff
@@ -22,14 +22,13 @@ namespace AZ
                     ->Version(6)
```
bumped several versions

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialInstanceHandler.h:
```diff
@@ -0,0 +1,72 @@
+#include <AzCore/Interface/Interface.h>
+#include <AzCore/std/smart_ptr/shared_ptr.h>
+
+namespace AZ
```
changed to a nested namespace in several files

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialShaderParameter.h:
```diff
@@ -0,0 +1,205 @@
+            AZStd::unordered_map<uint32_t, AZStd::vector<uint8_t>> m_structuredBufferData = {};
+            Data::Instance<RPI::ShaderResourceGroup> m_shaderResourceGroup = nullptr;
+            int m_materialTypeIndex;
+            int m_materialInstanceIndex;
```
added init

kh-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Public/Material/MaterialShaderParameter.h:
```diff
@@ -0,0 +1,205 @@
+            void SetStructuredBufferData(
+                const MaterialShaderParameterDescriptor* desc, const AZStd::span<const uint8_t>& deviceData, const int deviceIndex);
+
+            const MaterialShaderParameterLayout* m_layout;
```
done

kh-huawei commented on Gems/Atom/Asset/Shader/Code/AZSL/Platform/Mac/Metal/AzslcHeader.azsli:
```diff
@@ -17,6 +17,11 @@
 
+// use the Single Material SRG with maximum 32 textures per material
+// note: This assumes AZ_TRAIT_REGISTER_TEXTURES_PER_MATERIAL is defined via Atom_RPI_Traits_Platform.h
+#define AZ_TRAIT_FORCE_SINGLE_MATERIAL_SRG
```
the android header defines explicit bindless support (or at least the`UNBOUNDED_SIZE` implies that), so i assumed this works. I have yet to test that though.

moudgils replied:
UNBOUNDED_SIZE is used to simulate bindless in case native bindless support did not exist which is the case for metal although with metal it is a shader language issue and it can be resolved but this was a workaround if someone wanted to use bindless features on metal. For android UNBOUNDED_SIZE is set to null which implies that Android **can** support bindless via Vulkan drivers but it is up to the render pipeline to use it. The issue is that the bindless support is not well tested on the mobile platform and there is a good chance it is not mature across all hw/firmware versions. Hence we should refrain from making bindless a core requirement for Android until we do detailed testing across the top devices.

kh-huawei replied:
i've enabled single-material mode and non-bindless for android by default for now.

moudgils commented:
> I did some performance testing on my PC, with a RTX 3090, with two levels from the AutomatedTesting - project, where i added a camera-path and recorded about 60 seconds of rendering, with the values here averaged.
> 
> I did one test-run before refactoring, one after, and one with the AZ_TRAIT_FORCE_SINGLE_MATERIAL_SRG and AZ_TRAIT_REGISTER_TEXTURES_PER_MATERIAL both defined, with 32 textures per material-srg (called After, SM).
> 
> Also note that with DX12, i had VSync enabled by accident, which makes the CPU times not that useful, since it only shows we are <= 16.6 ms, but i didn't want to wait to recompile the engine again.
> 
> ## Level 10kEntityCpuPerfTest:
> A level with many instanced cubes, presumably all using the same material.
> 
> ### DX12
> * Before: 16.6 ms CPU (Vsync on), 5.11 ms GPU
> * After: 16.67 ms CPU (VSync on), 5.09 ms GPU
> * After, SM : 14.5 ms CPU (VSync off), 3.84 ms GPU
> 
> ### Vulkan
> * Before: 37.8 ms CPU, 6.57 ms GPU
> * After: 38.08 ms CPU, 6.78 ms GPU
> * After, SM: 38.46 ms CPU, 6.60 ms GPU
> 
> The Vulkan RHI doesn't really cope very well with this level in general: I had very bad spikes, with the CPU taking 8000 - 9000 ms sometimes, and the GPU 45 - 50 ms in all tests, with or without my changes.
> 
> ## Level PbrMaterialChartPerf:
> A level with two rows of spheres, with each sphere having slightly different PBR - settings (~20 spheres in total, seems like a difference in roughness mostly?)
> 
> ### DX12
> * Before: 16.6 ms CPU (Vsync on), 2.945 ms GPU
> * After: 8.35 ms CPU (VSync off), 2.09 ms GPU
> * After, SM: 8.23 ms CPU (VSync off), 1.45 ms GPU
> 
> ### Vulkan
> * Before: 1.96 ms CPU, 0.86 ms GPU
> * After: 2.05 ms CPU, 0.85 ms GPU
> * After, SM: 2.00 ms CPU, 0.83 ms GPU
> 
> ## Conclusion (for my Desktop GPU):
> CPU overhead seems to increases minimally with Vulkan. I unfortunately have no real data for DX12 on that, but i assume it's similar. GPU performance is between the old and the new implementation is very similar, but interestingly i see a noticeable improvement with the Single-Material mode. I'm not really clear why though: While i assume reading parameters from a constant buffer is slightly faster, that doesn't explain the performance of the old implementation. Maybe packing the Params in a struct allows for better compiler optimizations?

This looks good. Tagging @galibzon to add any data collected from Android related testing if possible.

moudgils commented on Gems/Atom/RPI/Assets/ShaderLib/Atom/RPI/ShaderResourceGroups/DefaultDrawSrg.azsli:
```diff
@@ -28,6 +28,8 @@ ShaderResourceGroup DrawSrg : SRG_PerDraw
 {
     // This SRG is unique per draw packet
     uint m_uvStreamTangentBitmask;
+    int m_materialTypeId;
```
Does this make more sense to be part of ObjectSRG?

kh-huawei replied:
good point. Moved to the ObjectSrg

kh-huawei commented on Gems/Atom/Feature/Common/Assets/ShaderLib/Atom/Features/Materials/SceneMaterialSrg.azsli:
```diff
@@ -0,0 +1,70 @@
+
+ShaderResourceGroup SceneMaterialSrg : SRG_PerMaterial
+{
+    Buffer<int> m_materialTypeBufferIndices;
```
Changed to ByteAddressBuffer.

kh-huawei commented:
I've did some rough testing on my personal android device (pixel 7a): I had to fix some asset serialization issues to get it to run, but my device is apparently modern enough to support bindless without issues.

The Results: 

13.5 FPS (~70 ms), MainPipeline
19.5 FPS (~51 ms), LowEndRenderPipeline
61.2 FPS (~16.4 ms) MobileRenderPipeline, vsync active
140 - 240 FPS (4 - 7 ms) MobileRenderPipeline, vsync disabled 

I've tested the same settings as before (without my changes, with my changes using bindless and SceneMaterialSrg, and using the SingleMaterialMode and bindless support disabled), but only for the PbrMaterialChartPerf - level. 
And my method this time was unfortunately just reading the performance numbers on the screen and guesstimating the average value, which produced effectively the same results for all settings.

I've also tried a brief test after i changed the typed `Buffer<int>` to a `ByteAddressBuffer` in the SceneMaterialSrg, but again, at least on my devices i didn't see a noticeable difference. 

Disabling vsync with the 'vsync_interval' CVAR was unreliable: It seemed to have an influence only if the app was suspended at least once. After that the frametime varied quite a lot for some time until settling towards ~6 ms.

moudgils reviewed (approved):
Approving this PR but highly recommend waiting for @galibzon to test and give the go-ahead after testing the changes on Android.

galibzon commented:
> Approving this PR but highly recommend waiting for @galibzon to test and give the go-ahead after testing the changes on Android.

I'm testing this now on Quest3 with OpenXRTest and our custom project.

galibzon commented:
I tested with OpenXRTest first, but unfortunately lots of important assets are not building properly. This is a compilation from scratch of the engine and assets using the branch of this PR:
[StandardPBR_materialtype.log](https://github.com/user-attachments/files/19983623/StandardPBR_materialtype.log)

galibzon reviewed (changes requested):
Please make sure that OpenXRTest at least compiles on Windows and all assets get processed successfully for both PC and android.

kh-huawei commented:
> Please make sure that OpenXRTest at least compiles on Windows and all assets get processed successfully for both PC and android.

This seems to mostly work for me out of the box:
- The project compiles (using clang & ninja on windows, currently testing MSVC)
- The assets for android get processed, but i get `float` to `half` - warnings from some shaders, although they seem unrelated to my changes
- I was able to start the XR_Office level, although using the MainPipeline. I didn't figure out how to simulate a VR headset, neither on my PC nor on my Phone...

Are you running into issues that my setup doesn't cover?

moudgils commented:
@galibzon Did you do asset build after doing a clean, i.e deleting the cache folder and building from scratch?

galibzon commented:
> @galibzon Did you do asset build after doing a clean, i.e deleting the cache folder and building from scratch?

Yes, indeed.

galibzon commented:
> > Please make sure that OpenXRTest at least compiles on Windows and all assets get processed successfully for both PC and android.
> 
> This seems to mostly work for me out of the box:
> 
>     * The project compiles (i tried both, clang 16 and vs2022, although only windows)
> 
>     * The assets for android get processed, but i get `float` to `half` - warnings from some shaders, although they seem unrelated to my changes
> 
>     * I was able to start the XR_Office level, although using the MainPipeline only. I didn't figure out how to simulate a VR headset, neither on my PC nor on my Phone...
> 
> 
> Are you running into issues that my setup doesn't cover?

I wonder if there's a missing file at your end that is not published in this PR. A few more commits have been added so I'll retry over the weekend. As I mentioned already, I did a Clean build with VS2022 on windows, and rebuilt all Assets from scratch when I reported the asset failures. Will try again.

moudgils commented:
> > > Please make sure that OpenXRTest at least compiles on Windows and all assets get processed successfully for both PC and android.
> > 
> > 
> > This seems to mostly work for me out of the box:
> > ```
> > * The project compiles (i tried both, clang 16 and vs2022, although only windows)
> > 
> > * The assets for android get processed, but i get `float` to `half` - warnings from some shaders, although they seem unrelated to my changes
> > 
> > * I was able to start the XR_Office level, although using the MainPipeline only. I didn't figure out how to simulate a VR headset, neither on my PC nor on my Phone...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Are you running into issues that my setup doesn't cover?
> 
> I wonder if there's a missing file at your end that is not published in this PR. A few more commits have been added so I'll retry over the weekend. As I mentioned already, I did a Clean build with VS2022 on windows, and rebuilt all Assets from scratch when I reported the asset failures. Will try again.

Thanks Galib. I also recommended to @kh-huawei to do a clean build  (from synching to just the changes in this Pr) on his side and re-post the end result with re-pro steps.

kh-huawei commented:
> I wonder if there's a missing file at your end that is not published in this PR. A few more commits have been added so I'll retry over the weekend. As I mentioned already, I did a Clean build with VS2022 on windows, and rebuilt all Assets from scratch when I reported the asset failures. Will try again.

The new commits were mostly from rebasing onto the new development branch, i don't think they fixed anything there. I am also using the refactored materials in one other engine instance on my PC, so i should have noticed missing files there? Although i am using the commits from the last refactoring PR there, but still...

I've tried now to get the OpenXRTest - project to run on my Pixel 7a (Following your "OpenXR Developer Workflow: Part 1 - Setup" Video for the most part, except i skipped the MetaQuest - specific things), and i managed to get the XR_Office to load: I eventually ran into an error on the device ("Sampler Input index '-1' out of range [0,0)."), which i am currently looking into, but i did not run into any asset processing issues.
I've also tried to build in 'release' now, since i usually use 'profile', and that also revealed two unused variables, but i did not run into any Asset Processing issues. In fact i didn't even get any assets with warnings, which is a bit suspicious too, well.
I've documented my steps here: [compile_for_android.txt](https://github.com/user-attachments/files/20123017/compile_for_android.txt), maybe it's useful to figure out where we differ.

One suspicion i still have is that this is related to bundled (or packaged?) assets somehow, and the Asset-Processor operates slightly different then? Although i have no real experience in that regard, unfortunately...

galibzon commented:
I got latest today of this branch, and was able to validate OpenXRTest (XR_Office) on Windows11 and Quest3. The Asset Processing errors are gone, except for:
![image](https://github.com/user-attachments/assets/9dbe2a35-e796-4531-a487-af47d4b3eba8)


**The Problem**: It seems Materials can't find textures on Quest3:
![image](https://github.com/user-attachments/assets/29d90e28-a02c-4a5a-acaa-298ca9ea8da8)

The Editor looks fine:
![image](https://github.com/user-attachments/assets/82d50ae3-7853-4a10-a02b-f1c223034378)

kh-huawei commented:
I was able to reproduce this with the mobile-pipeline running on my PC: The defines selecting the non-bindless version weren't properly propagated.
This should be fixed now, and the screenshot from my Pixel 7 currently looks like this:
![openXR_android](https://github.com/user-attachments/assets/159b2b2d-a0b7-42e4-8917-9ef05c958137)

galibzon commented:
> Thanks for fixing this. I am ok with merging this change. But we need to broadcast this change to everyone on discord as an #impactful change. @kh-huawei feel free to broadcast now while we wait for @galibzon to approve the PR.

I'll test this again.

moudgils commented:
This PR looks good to merge. Just ping the appropriate channels on discord once this is merged. Thanks for all the work :)
